xocto/ranges.py

```diff
@@ -891,10 +903,12 @@ def union(self, other: Range[datetime.datetime]) -> Optional["FiniteDatetimeRang
         if base_union is None:
             return None
 
-        assert base_union.boundaries == RangeBoundaries.INCLUSIVE_EXCLUSIVE
-        assert base_union.start is not None
-        assert base_union.end is not None
-        return FiniteDatetimeRange(base_union.start, base_union.end)
+        if _is_half_finite_datetime_range(other):
+            return HalfFiniteDatetimeRange(
```

Comment:
I'm deliberately calling the constructor here rather than using `cast` (we use `cast` in some other places in `xocto/ranges`) - see https://github.com/octoenergy/xocto/issues/193